Srv6 default vrf

pguibert6WIND opened the pull request:
default VRF should also be able to import/export srv6 routes.
Do the necessary changes.

pguibert6WIND committed:
bgpd: explain that 'vpn_leak_postchange_all()' is also called by SRv6

The 'vpn_leak_postchange_all()' comment indicates that the function
is called only upon bgp instance creation.
Since now, SRv6 calls it, let us update the comment.

Fixes: a0281b2eab10 ("bgpd: cli for srv6-locator assignment (step4)")
Fixes: d79ff732cdcd ("bgpd: handle srv6 locator notification and update vpn-rib")
Fixes: 0249b8b61206 ("bgpd: add no-cli for srv6 on bgpd-side")
Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
bgpd: notify bgp updates from default VRF upon SRv6 changes

An L3VPN network can be configured on the main BGP instance,
with an SRv6 SID. By declaring a network, the network is
exported, and a VPNv6 update is created, but is not selected.

The below BGP VPN configuration in the default VRF has been
used:

> router bgp 1
>  segment-routing srv6
>   locator loc1
>  exit
>  !
>  address-family ipv6 unicast
>   sid vpn export auto
>   rd vpn export 1:30
>   rt vpn both 77:77
>   import vpn
>   export vpn
>   network 2001:7::/64
>  exit-address-family

The below BGP VPNv6 update is created, but is not selected.
Also, no SRv6 SID is allocated.

> # show bgp ipv6 vpn
> [..]
>    Network          Next Hop            Metric LocPrf Weight Path
> Route Distinguisher: 1:30
>    2001:7::/64      ::@0<                    0         32768 ?
> # show bgp segment-routing srv6
> [..]
> bgps:
> - name: default
>   vpn_policy[AFI_IP].tovpn_sid: (null)
>   vpn_policy[AFI_IP6].tovpn_sid: (null)
>   per-vrf tovpn_sid: (null)
>

The SID in the main instance has not been updated, whereas
in other BGP VRF instances, a SID is available. The update
happens in the 'bgp_vpn_leak_postchange_all()' function.
The main BGP instance is not parsed, so let us refresh the
L3VPN updates of the default BGP instance too.

This results in having a selected BGP VPNv6 update, with a
SID.

> # show bgp segment-routing srv6
> [..]
> bgps:
> - name: default
>   vpn_policy[AFI_IP].tovpn_sid: (null)
>   vpn_policy[AFI_IP6].tovpn_sid: 2001:db8:1:1:100::
>   per-vrf tovpn_sid: (null)

Fixes: 3bd70bf8f37b ("bgpd: fix NULL dereference in vrf-vpn leak config if before default instance")

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
bgpd: fix L3VPN cases on the default VRF

The BGP L3VPN code imports and exports IPv4/IPv6 prefixes
from the BGP main instance and from the BGP VRF instances.
However, in some scenarios, the BGP L3VPN routes may not be
refreshed properly.

The first change in the bgp_vpn_leak_unimport() function
is related to unimporting route entries when deleting a
BGP instance.

The second change in the is_bgp_vrf_mplsvpn() function is
is related to the visibility of the L3VPN MIB applied to
the default VRF instance too.

Fixes: ("9ecf931b13cf bgpd: no router bgp cleanup vrf leaked vpn routes")
Fixes: 2fceba1fa7de ("bgpd: add utility to check if a vrf is MPLS VPN")
Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
bgpd: fix default VRF L3VPN changes, when unsetting SRv6 locator

When unsetting the default srv6 locator of the main BGP instance,
all the SIDs are unallocated, except the one from the main BGP
instance.

The below configuration uses a locator and request a SID for
the main BGP instance:

> router bgp 1
>  segment-routing srv6
>   locator loc1
>  exit
>  !
>  address-family ipv6 unicast
>  redistribute connected
>   sid vpn export auto
>   rd vpn export 1:30
>   rt vpn both 77:77
>   export vpn
>   import vpn
>  exit-address-family

An SID is allocated for the prefixes imported from the
ipv6 address family. But the SID is not unallocated,
when unconfiguring the locator:

> router bgp 1
> no segment-routing srv6
> do show bgp segment-routing srv6
> bgps:
> - name: default
>   vpn_policy[AFI_IP].tovpn_sid: (null)
>   vpn_policy[AFI_IP6].tovpn_sid: 2001:db8:1:1:100::
>   per-vrf tovpn_sid: (null)

The default BGP instance is ignored when the locator
is unconfigured. Add the default BGP instance to check
if any SIDs should be unallocated or not.

Fixes: 0249b8b61206 ("bgpd: add no-cli for srv6 on bgpd-side")
Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND commented:
ci:rerun

riw777 reviewed (approved):
looks good

riw777 commented:
ci errors in ospf setup ... don't seem related ... trying again
ci:rerun

pguibert6WIND commented:
> Do you have any insight into why we were skipping the default VRF prior to these changes? It looks like these commits are just making it so that we no longer skip the default VRF - I don't see anything indicating why those checks were in place to begin with or whether this might tickle a known issue.


I have no idea why default vrf was processed separately, though I intensively used it in L3VPN tests with MPLS. I just noticed this issue while playing with srv6.

> 
> I think we'll at least want a topotest that does validation before/after configuring + unconfiguring SRv6 export/import in the default VRF -- ideally testing that other tenant VRFs w/ SRv6 already configured aren't disrupted by any dependency on the default VRF that might now get exposed.

I will do it later by appending changes to the `bgp_srv6l3vpn_over_ipv6` test.

dmytroshytyi committed:
bgpd: fix the bgp->name null pointer

Avoid the crash provoked by bgp->name null pointer in snmp context.
If bgp->name == NULL, replace it by "defalut" vrf name.

Signed-off-by: Dmytro Shytyi <dmytro.shytyi@6wind.com>

dmytroshytyi committed:
bgpd: snmp, fix possible issues when default vrf is not set.

When importing VPN updates in the default VRF, a crash happens
when dumping the L3VPN MIB. To illustrate, the following crash log example is displayed:

BGP[7301]: Received signal 11 at 1684255316 (si_addr 0x0, PC 0x7fa09b992b05); aborting...
BGP[7301]: /lib/x86_64-linux-gnu/libfrr.so.0(zlog_backtrace_sigsafe+0xb3) [0x7fa09bbd93f2]
BGP[7301]: /lib/x86_64-linux-gnu/libfrr.so.0(zlog_signal+0x1b4) [0x7fa09bbd9257]
BGP[7301]: /lib/x86_64-linux-gnu/libfrr.so.0(+0x10568d) [0x7fa09bc1c68d]
BGP[7301]: /lib/x86_64-linux-gnu/libc.so.6(+0x42520) [0x7fa09b837520]
BGP[7301]: /lib/x86_64-linux-gnu/libc.so.6(+0x19db05) [0x7fa09b992b05]
BGP[7301]: /usr/lib/x86_64-linux-gnu/frr/modules/bgpd_snmp.so(+0x62ff) [0x7fa09b5f42ff]
BGP[7301]: /usr/lib/x86_64-linux-gnu/frr/modules/bgpd_snmp.so(+0x647c) [0x7fa09b5f447c]
BGP[7301]: /usr/bin/bgpd(+0x1e1aab) [0x55b1ce417aab]
BGP[7301]: /usr/bin/bgpd(+0x1e261b) [0x55b1ce41861b]
BGP[7301]: /lib/x86_64-linux-gnu/libfrr.so.0(if_up_via_zapi+0x2c) [0x7fa09bbbe43c]
BGP[7301]: /lib/x86_64-linux-gnu/libfrr.so.0(+0x13df63) [0x7fa09bc54f63]
BGP[7301]: /lib/x86_64-linux-gnu/libfrr.so.0(+0x14267e) [0x7fa09bc5967e]
BGP[7301]: /lib/x86_64-linux-gnu/libfrr.so.0(thread_call+0x1c2) [0x7fa09bc35fcc]
BGP[7301]: /lib/x86_64-linux-gnu/libfrr.so.0(frr_run+0x257) [0x7fa09bbcd6e8]
BGP[7301]: /usr/bin/bgpd(main+0x4f4) [0x55b1ce2ef98c]
BGP[7301]: /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7fa09b81ed90]
BGP[7301]: /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80) [0x7fa09b81ee40]
BGP[7301]: /usr/bin/bgpd(_start+0x25) [0x55b1ce2edae5]
BGP[7301]: in thread zclient_read scheduled from /build/make-pkg/output/_packages/cp-routing/src/lib/zclient.c:4397 zclient_event()

The 'name' attribute of the BGP instance is attempted to be accessed,
in multiple functions.

Fix this by using the "default" vrf name, which is returned by the
VRF_DEFAULT_NAME macro, wherever the default BGP instance may be
used.

Signed-off-by: Dmytro Shytyi <dmytro.shytyi@6wind.com>
Signed-off-by: Philippe guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
lib: fix next vrf interface lookup avoid loopback interface

When the default vrf is submitted, the returned interface
may be the loopback interface, which may not be valid.
Actually, on vrfs, the vrf interface is bypassed. When
being on default vrf, the 'lo' interface should be
bypassed too.
Fix this by avoiding returning the lo interface too.

Fixes: 0760a74d2fc2 ("lib: add utility to get the next index in a vrf")

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
topotests: add mpls l3vpn test with default vrf

There is no MPLS VPN test with default VRF.
Add a test to ensure that the MPLS VPN MIB also works
with the default VRF.
An additional check is done to refine the real number
of interfaces based on the presence or not of the gre0,
gretap, and erspan interfaces.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>
Signed-off-by: Louis Scalbert <louis.scalbert@6wind.com>

riw777 reviewed (approved):
still looks good ... failing ci, though

riw777 commented:
ci:rerun

riw777 commented:
looks like this is causing a failure in mpls topo tests ... probably needs to be looked at

riw777 commented:
maybe this should be rebased because it's pretty old to see if it can pass ci ... let me try and see if this works ...

@Mergifyio rebase

riw777 commented:
is anyone still looking at this?

ton31337 commented on bgpd/bgp_mplsvpn_snmp.c:
```diff
@@ -554,16 +554,23 @@ static bool is_bgp_vrf_active(struct bgp *bgp)
 	if (vrf == NULL)
 		return false;
+
+	if (bgp->name)
```
If the default vrf is used, then bgp->name is NULL? If so, this is used in many places and could be a macro/inline function/etc.?

pguibert6WIND replied:
I am not so sure bgp->name can be null each time default vrf is used.
I prefer to separate the problems.

ton31337 commented:
Could you add a small documentation/example on how to use srv6 on default vrf also? This would be useful for the end-users and overall to understand on how to use it.

mwinter-osr commented:
CI:rerun Retest for new ASAN Memory Leaks

riw777 commented:
is anyone still looking at this? @pguibert6WIND ?

riw777 commented:
@pguibert6WIND we're just waiting on the doc @ton31337 asked for above?

pguibert6WIND commented:
> @pguibert6WIND we're just waiting on the doc @ton31337 asked for above?

@riw777 , i don't see the point in having a doc for default vrf, we did not do this for l3vpn mpls. I don't understand why we would need to do it for srv6..

riw777 commented:
still some lint errors ... :-)

ton31337 commented on tests/topotests/bgp_snmp_mplsl3vpn/r1/bgpd.conf:
```diff
@@ -6,6 +6,15 @@ router bgp 65000
  neighbor 10.4.4.4 timers connect 10
  !
+ address-family ipv4 unicast
+  network 8.8.8.8/32
```
No Google.

ton31337 reviewed (dismissed):
Please apply formatting.

pguibert6WIND commented:
> Please apply formatting.

done

louis-6wind commented on tests/topotests/bgp_snmp_mplsl3vpn/test_bgp_snmp_mplsvpn.py:
```diff
@@ -253,19 +253,19 @@ def _convergence():
-    "mplsL3VpnConnectedInterfaces": "3",
+    "mplsL3VpnConfiguredVrfs": "3",
+    "mplsL3VpnActiveVrfs": "3",
+    "mplsL3VpnConnectedInterfaces": "10",
```
There is not always gre0, gretap, erspan interfaces.

When there is no gre0, gretap, erspan, the number is 7

pguibert6WIND replied:
add a patch to account for the real number of interfaces in the default vrf.

riw777 reviewed (approved):
looks good

riw777 commented:
I'm not certain who has blocking changes on this one ... we should clear it up so we can get this fixed up and pushed ...

riw777 commented:
@ton31337 can we get the blocker cleared on this?

pguibert6WIND commented:
> @pguibert6WIND can you clear the conflicts? We still need to get the blocker cleared so this can go in ...

done

ton31337 commented:
Seems we have some crashes here.

pguibert6WIND commented:
ci:rerun not reproduced locally

riw777 commented:
rerunning just the failed test ... maybe this just needs to rebased to clear this problem? the version is pretty old at this point ...

riw777 commented:
@Mergifyio rebase

riw777 commented:
I don't think this should be related, but ci fails here:

```E   AssertionError: mplsL3VpnConnectedInterfaces should be 10: value 16
   assert False
    +  where False = test_oid('mplsL3VpnConnectedInterfaces', '10')
    +    where test_oid = <lib.snmptest.SnmpTester object at 0xb62ff190>.test_oid```

I've tried rerunning this test a couple of times ...

ben-raskin commented:
I'm just curious, is this PR still being worked on. No pressure, I would just find this feature useful. If I can help in anyway to get this PR merged in lmk

pguibert6WIND commented:
Hi @ben-raskin , there are issues with SNMP L3VPN MIB usage, if you have some time to understand what happens. I suspect the issue partially resides in the fact that some default interfaces (ipip0 for instance) are not present in the environments used. (or vice-versa)

ben-raskin commented:
> Hi @ben-raskin , there are issues with SNMP L3VPN MIB usage, if you have some time to understand what happens. I suspect the issue partially resides in the fact that some default interfaces (ipip0 for instance) are not present in the environments used. (or vice-versa)

Yeah I don't see any changes in bgp_mplsvpn_snmp.c that could break something. I noticed in the bgpd config for r1 you didn't include segment-routing section (https://docs.frrouting.org/en/latest/bgp.html#l3vpn-srv6). You might need to include it to create SRv6 route endpoints.

pguibert6WIND committed:
topotests: bgp_srv6l3vpn, add l3vpn case using default vrf

Add l3vpn srv6 use case which uses default vrf.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
bgpd: accept sid and locator notifications on default vrf

Accept SID and Locator notifications on the default VRF.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
bgpd: fix remove control on bgp vrf instance on bgp_lookup_by_rd()

Some function apply filtering on events that are only on vrf instances.
For SRv6, we can have events on default vrf instance too. Remove that
control.

Fixes: 46dbf9d0c0b9 ("bgpd: Implement ACCEPT_OWN extended community")

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND committed:
bgpd: fix crash when running bgp_rfapi_basic_sanity test

The following crash has happened.

> #0  __pthread_kill_implementation (no_tid=0, signo=11, threadid=134363256068544) at ./nptl/pthread_kill.c:44
> #1  __pthread_kill_internal (signo=11, threadid=134363256068544) at ./nptl/pthread_kill.c:78
> #2  __GI___pthread_kill (threadid=134363256068544, signo=signo@entry=11) at ./nptl/pthread_kill.c:89
> #3  0x00007a33e1c42476 in __GI_raise (sig=sig@entry=11) at ../sysdeps/posix/raise.c:26
> #4  0x00007a33e210feca in core_handler (signo=11, siginfo=0x7ffd4a405130, context=<optimized out>) at lib/sigevent.c:268
> #5  <signal handler called>
> #6  vpn_leak_to_vrf_update_onevrf (to_bgp=0x5a70a945d9b0, from_bgp=0x5a70a945d9b0, path_vpn=0x5a70a9488b50, prd=0x0, from=0x5a70a9482350) at bgpd/bgp_mplsvpn.c:2256
> #7  0x00005a7093ca627c in vpn_leak_to_vrf_update_all (to_bgp=to_bgp@entry=0x5a70a945d9b0, vpn_from=vpn_from@entry=0x5a70a945d9b0, afi=afi@entry=AFI_IP) at bgpd/bgp_mplsvpn.c:2839
> #8  0x00005a7093ca789d in vpn_leak_postchange (bgp_vrf=0x5a70a945d9b0, bgp_vpn=0x5a70a945d9b0, afi=AFI_IP, direction=BGP_VPN_POLICY_DIR_FROMVPN) at ./bgpd/bgp_mplsvpn.h:270
> #9  vpn_leak_postchange_all () at bgpd/bgp_mplsvpn.c:4001
> #10 0x00007a33e2123366 in event_call (thread=thread@entry=0x7ffd4a405a50) at lib/event.c:2005
> #11 0x00007a33e20c3828 in frr_run (loop=0x5a70a8c0ff90) at lib/libfrr.c:1252
> #12 0x00005a7093c45122 in main (argc=<optimized out>, argv=<optimized out>) at bgpd/bgp_main.c:565

prefix attached to pdest is null. Protect it.

Signed-off-by: Philippe Guibert <philippe.guibert@6wind.com>

pguibert6WIND commented:
ci:rerun

pguibert6WIND commented:
ci:rerun

riw777 commented:
@frrbot autoclose in 2 weeks
